### PR TITLE
Cluster lifecycle web ui

### PR DESCRIPTION
**Pull request by bzapico**
#### What does this PR do?
Implements the[ mockup design for the cluster lifecycle management](https://xd.adobe.com/spec/16138a00-1a69-42ff-4ba4-1c39bc50866b-6022/)

The current scope of this task includes:
- Modal windows with states. (Applications and resources view)
- A new button in the graph view to show the modal. (Applications and resources view)
- Contextual action buttons → Static contextual menu with those options: Edit, cordon, uncordon, more info(got to the existing view for now) and drain. We should show an error notification when it’s necessary. (Resources view)
- The removal of instances column and the new column with status color. (Resources view)
- Add the new colors to the graph
- Creation of tools service

#### Where should the reviewer start?
- tools.component.ts
- tools.service.ts
- applications and resources components
- backend.service.ts

#### What is missing?
The typo in the cluster status enum should change when task [NP-2448](NP-2448) will be done

#### How should this be manually tested?
1. Run the app
2. Go to Resources view
3. Verify the performance of the contextual menu in list card
4. Verify the performance of the graph and search box in the graph
5. Click on the right bottom button in the graph to see the status legend
6. Go to Applications view
7. Verify that everything runs ok and all the states are shown

#### What are the relevant tickets?

- [NP-1929](https://nalej.atlassian.net/browse/NP-1929)

#### Screenshots (if appropriate)
Captura de pantalla 2019-12-02 a las 16 43 57

### CODE REVIEW

File: src/app/tools/tools.component.ts [hunk 1]
```diff
@@ -225,7 +293,7 @@ export class ToolsComponent implements OnInit {
       for (let indexInstance = 0, instancesLength = this.instances.length; indexInstance < instancesLength; indexInstance++) {
         if (this.areIncludedInstancesWithError
             && (this.instances[indexInstance].status_name.toLowerCase() === AppStatus.Error
-                || this.instances[indexInstance].status_name.toLowerCase() === AppStatus.DeploymentError)) {
+                || this.instances[indexInstance].status_name.toLowerCase() === AppStatus.Deployment_error)) {
```
Comment (jrodva) [inline src/app/tools/tools.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/tools/tools.component.ts [hunk 2]
```diff
@@ -167,33 +205,63 @@ export class ToolsComponent implements OnInit {
...
+   * Return an specific border color depending on the node status
+   * @param status Status name
+   */
+  getBorderColor(status: string): string {
+    switch (status.toLowerCase()) {
+      case ClusterStatus.Online_cordon:
+        return ToolsComponent.STATUS_BORDER_COLORS.ONLINE_CORDON;
+      case ClusterStatus.Offline_cordon:
+        return ToolsComponent.STATUS_BORDER_COLORS.OFFLINE_CORDON;
+      case ClusterStatus.Unknown:
```
Comment (jrodva) [inline src/app/tools/tools.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/tools/tools.component.ts [hunk 3]
```diff
@@ -167,33 +205,63 @@ export class ToolsComponent implements OnInit {
...
-      case ClusterStatus.Provisioned:
-      case ClusterStatus.Installing:
+      case ClusterStatus.Scaling:
       case ClusterStatus.Uninstalling:
-      case ClusterStatus.Decommissioning:
-      case AppStatus.Queued:
-      case AppStatus.Deploying:
+      case ClusterStatus.Online_cordon:
+      case ClusterStatus.Install_in_progress:
+      case ClusterStatus.Offline_cordon:
```
Comment (jrodva) [inline src/app/tools/tools.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/tools/tools.component.ts [hunk 4]
```diff
@@ -167,33 +205,63 @@ export class ToolsComponent implements OnInit {
...
+      case AppStatus.Running:
       case ClusterStatus.Offline:
-      case ClusterStatus.OfflineCordon:
+      case AppStatus.Terminated:
+      case ClusterStatus.Decomisioning:
+      case AppStatus.Terminating:
+      case ClusterStatus.Failure:
+      case AppStatus.Deployment_error:
+      case AppStatus.Incomplete:
+      case AppStatus.Planning_error:
```
Comment (jrodva) [inline src/app/tools/tools.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/applications/applications.component.ts [hunk 5]
```diff
@@ -767,7 +791,7 @@ export class ApplicationsComponent extends ToolsComponent implements OnInit, OnD
       if (this.filters.clusters && this.filters.instances) {
         if ((this.areIncludedInstancesWithError
             && instance.status_name.toLowerCase() !== AppStatus.Error
-            && instance.status_name.toLowerCase() !== AppStatus.DeploymentError)
+            && instance.status_name.toLowerCase() !== AppStatus.Deployment_error)
```
Comment (jrodva) [inline src/app/applications/applications.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/definitions/enums/app-status.enum.ts [hunk 6]
```diff
@@ -12,16 +12,15 @@
...
-    Running = 'running',
     Scheduled = 'scheduled',
-    Queued = 'queued',
-    Deploying = 'deploying'
+    Deploying = 'deploying',
+    Running = 'running',
+    Terminating = 'terminating',
+    Incomplete = 'incomplete',
+    Planning_error = 'planning_error',
+    Deployment_error = 'deployment_error',
```
Comment (jrodva) [inline src/app/definitions/enums/app-status.enum.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/definitions/enums/cluster-status.enum.ts [hunk 7]
```diff
@@ -12,19 +12,17 @@
...
-    Running = 'running',
-    Unknown = 'unknown',
     Provisioning = 'provisioning',
     Provisioned = 'provisioned',
-    Installing = 'installing',
-    Installed = 'installed',
+    Install_in_progress = 'install_in_progress',
+    Online = 'online',
+    Online_cordon = 'online_cordon',
+    Offline_cordon = 'offline_cordon',
```
Comment (jrodva) [inline src/app/definitions/enums/cluster-status.enum.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/definitions/enums/cluster-status.enum.ts [hunk 8]
```diff
@@ -12,19 +12,17 @@
...
-    OfflineCordon = 'offline_cordon',
-    Error = 'error',
-    // STATE
-    Running = 'running',
-    Unknown = 'unknown',
     Provisioning = 'provisioning',
     Provisioned = 'provisioned',
-    Installing = 'installing',
-    Installed = 'installed',
+    Install_in_progress = 'install_in_progress',
```
Comment (jrodva) [inline src/app/definitions/enums/cluster-status.enum.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/resources/cluster-status-info/cluster-status-info.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (jrodva) [inline src/app/resources/cluster-status-info/cluster-status-info.component.spec.ts]: @bzapico we should add the apache headers.

File: src/app/resources/cluster-status-info/cluster-status-info.component.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * Models that hold the cluster and instances enums
+   */
+  clusterStatusList = ClusterStatus;
+  appStatusList = AppStatus;
+  /**
+   * Method that returns an array of a given clusters and instances' enums property
+   */
+  keys = Object.keys;
+  constructor(
```
Comment (jrodva) [inline src/app/resources/cluster-status-info/cluster-status-info.component.ts]: Please add a line before the constructor to maintains the current format.

File: src/app/resources/resources.component.ts [hunk 11]
```diff
@@ -209,6 +224,18 @@ export class ResourcesComponent extends ToolsComponent implements OnInit, OnDest
...
+   * Opens the modal view that holds the edit cluster component
+   */
+  openStatusCluster() {
+    const initialState = {
+      organizationId: this.organizationId
+    };
+    this.modalRef = this.modalService.show(ClusterStatusInfoComponent, { initialState, backdrop: 'static', ignoreBackdropClick: false });
+    this.modalRef.content.closeBtnName = 'Close';
+    this.modalService.onHide.subscribe((reason: string) => {
+    });
```
Comment (jrodva) [inline src/app/resources/resources.component.ts]: @bzapico this block is unnecesary

File: src/app/resources/resources.component.ts [hunk 12]
```diff
@@ -391,10 +509,11 @@ export class ResourcesComponent extends ToolsComponent implements OnInit, OnDest
     let online = 0;
     if (clusters) {
       clusters.forEach(cluster => {
-        if (cluster.status_name === 'ONLINE') {
+        if (cluster.status_name === 'ONLINE' || 'ONLINE_CORDON') {
           online += 1;
```
Comment (jrodva) [inline src/app/resources/resources.component.ts]: It's better to use online++

File: src/app/resources/resources.component.ts [hunk 13]
```diff
@@ -451,7 +570,7 @@ export class ResourcesComponent extends ToolsComponent implements OnInit, OnDest
         }
         if ((this.areIncludedInstancesWithError
             && instance.status_name.toLowerCase() !== AppStatus.Error
-            && instance.status_name.toLowerCase() !== AppStatus.DeploymentError)
+            && instance.status_name.toLowerCase() !== AppStatus.Deployment_error)
```
Comment (jrodva) [inline src/app/resources/resources.component.ts]: @bzapico Enums have to be written using PascalCase http://wiki.c2.com/?PascalCase

File: src/app/tools/tools.component.ts [hunk 14]
```diff
@@ -37,24 +37,52 @@ export class ToolsComponent implements OnInit {
   /**
    * Refresh ratio
    */
-  static readonly REFRESH_INTERVAL = 20000;
+  static readonly REFRESH_INTERVAL = 2000;
```
Comment (jrodva) [inline src/app/tools/tools.component.ts]: @bzapico we should use 20000.

File: src/app/resources/cluster-status-info/cluster-status-info.component.ts (hunk 10 above)
Comment (bzapico) [reply src/app/resources/cluster-status-info/cluster-status-info.component.ts]: Done

File: src/app/resources/cluster-status-info/cluster-status-info.component.spec.ts (hunk 9 above)
Comment (bzapico) [reply src/app/resources/cluster-status-info/cluster-status-info.component.spec.ts]: Done

File: src/app/resources/resources.component.ts (hunk 11 above)
Comment (bzapico) [reply src/app/resources/resources.component.ts]: Done

File: src/app/tools/tools.component.ts (hunk 14 above)
Comment (bzapico) [reply src/app/tools/tools.component.ts]: Done

File: src/app/resources/resources.component.ts (hunk 12 above)
Comment (bzapico) [reply src/app/resources/resources.component.ts]: Done

File: src/app/tools/tools.component.ts (hunk 2 above)
Comment (bzapico) [reply src/app/tools/tools.component.ts]: Done

File: src/app/tools/tools.component.ts (hunk 3 above)
Comment (bzapico) [reply src/app/tools/tools.component.ts]: Done

File: src/app/tools/tools.component.ts (hunk 4 above)
Comment (bzapico) [reply src/app/tools/tools.component.ts]: Done

File: src/app/applications/applications.component.ts (hunk 5 above)
Comment (bzapico) [reply src/app/applications/applications.component.ts]: Done

File: src/app/definitions/enums/app-status.enum.ts (hunk 6 above)
Comment (bzapico) [reply src/app/definitions/enums/app-status.enum.ts]: Done

File: src/app/definitions/enums/cluster-status.enum.ts (hunk 7 above)
Comment (bzapico) [reply src/app/definitions/enums/cluster-status.enum.ts]: Done

File: src/app/definitions/enums/cluster-status.enum.ts (hunk 8 above)
Comment (bzapico) [reply src/app/definitions/enums/cluster-status.enum.ts]: Done

File: src/app/resources/resources.component.ts (hunk 13 above)
Comment (bzapico) [reply src/app/resources/resources.component.ts]: Done

**Comment by jrodva** [comment]
Please check the status info window
image

**Pull request merged**